R/s3.R

```diff
@@ -29,95 +31,135 @@ s3_config <- function(path, bucket, profile='default', config_file=getOption("sc
+#' @param mock_get character. if remote_ind and local_source imply different
+#'   local file locations, should the current local file (implied by
+#'   local_source) be left alone ('none'), or copied ('copy') or moved ('move')
+#'   to the location implied by remote_ind? If 'copy' or 'move' are used, and if
+#'   gd_get will be called in an upcoming command, this argument may help to
+#'   avoid an unnecessary download from S3 back to this computer
+#'   because `s3_get` skips the download if there's already a local file in the
+#'   right place with the right contents (MD5 hash).
+#' @param on_exists what to do if the file already exists - replace, or
+#'   throw an error? the default is to replace
+#' @param verbose logical Used in this function and passed through aws.s3::put_object to
+#'   aws.s3::s3HTTP
+#' @param config_file character name of the YAML file containing
+#'   project-specific configuration information for S3
 #' @param ind_ext the indicator file extension to expect at the end of
 #'   remote_ind
 #' @export
-s3_put <- function(
-  data_file, config_file=getOption("scipiper.s3_config_file"),
-  ind_ext=getOption("scipiper.ind_ext")) {
+s3_put <- function(remote_ind, local_source,  mock_get=c('copy','move','none'),
+                   on_exists=c('replace','stop'), verbose = FALSE,
+                  config_file=getOption("scipiper.s3_config_file"),
+                  ind_ext=getOption("scipiper.ind_ext")) {
   
-  warning('s3_put is out of date relative to gd_put') # need to update s3_xx with lessons learned from gd_xx
+  # check arguments
+  mock_get <- match.arg(mock_get)
   
-  require_libs('aws.signature', 'aws.s3')
-  ind_file <- as_ind_file(data_file, ind_ext=ind_ext)
+  # decide whether local_source is an indicator or data file and find the data file 
+  # if it is an indicator
+  local_file <- check_local_source(local_source, ind_ext)
   
-  # post the file from the local data_file to S3
+  # identify the remote data file to be indicated by remote_ind
+  data_file <- as_data_file(remote_ind, ind_ext=ind_ext)
+  
+  # prepare to use S3
+  require_libs('aws.signature', 'aws.s3')
   s3_config <- yaml::yaml.load_file(config_file)
-  message("Uploading ", data_file, " to S3")
   aws.signature::use_credentials(profile = s3_config$profile)
-  key <- file.path(s3_config$path, basename(data_file))
-  success <- aws.s3::put_object(
-    file = data_file,
-    object = key, 
-    bucket = s3_config$bucket)
+  
+  # determine whether and where the remote file exists
+  bucket_contents <- aws.s3::get_bucket_df(bucket = s3_config$bucket)
+  exists_on_s3 <- local_file %in% bucket_contents$Key
+  
+  #upload to S3 - note that S3 is a flat file system, so folders don't need
+  #to be created.  'Directories' are just part of object keys
+  match.arg(on_exists)
+  if(exists_on_s3 && on_exists == "stop") {
+    stop('File already exists and on_exists==stop')
+  } else {
+    if(verbose) message("Uploading ", local_file, " to S3")
```

Comment:
even though the `aws.s3::put_object` call is the same for both a fresh upload and a replacement (at least, i assume so based on your code - is that right?), it would be nice to have the message distinguish between those two options. could you have this message choose between "Uploading" and "Replacing" to indicate which one is happening?